src/UI/Art/SpriteArtist/OverlayArtist.java

```diff
@@ -0,0 +1,110 @@
+package UI.Art.SpriteArtist;
+
+import java.awt.AlphaComposite;
+import java.awt.Color;
+import java.awt.Composite;
+import java.awt.Graphics;
+import java.awt.Graphics2D;
+import java.awt.image.BufferedImage;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import Engine.XYCoord;
+import Terrain.GameMap;
+import UI.GameOverlay;
+
+public class OverlayArtist
+{
+  private static final Color HIGHLIGHT_COLOR = new Color(255, 255, 255, 160); // white
+
+  public static final Color MOBILE_FIRE_EDGE = new Color(240,   0,   0, 255); // red
+  public static final Color SIEGE_FIRE_EDGE  = new Color(  0,   0, 160, 255); // dark blue
+  public static final Color FIRE_FILL        = new Color(  0,   0,   0,   0);
+
+  private static final int OVERLAY_EDGE_THICKNESS = 2;
+
+  static AlphaComposite buffComposite = null;
+  static long lastCompositeCreationTime = 0;
+
+  public static void drawHighlights(Graphics bigG,
+                                    GameMap gameMap,
+                                    Collection<GameOverlay> inputOverlays,
+                                    int drawX, int drawY,
+                                    int viewWidth, int viewHeight,
+                                    int tileSize)
+  {
+    if( viewWidth < 1 || viewHeight < 1 )
+      return;
+
+    BufferedImage overlayImage = SpriteLibrary.createTransparentSprite(viewWidth, viewHeight);
+    Graphics og = overlayImage.getGraphics();
+
+    // Only show stuff we're allowed to
+    Collection<GameOverlay> overlays = new ArrayList<GameOverlay>();
+    for( GameOverlay ov : inputOverlays )
+      if( null == ov.origin || !gameMap.isLocationFogged(ov.origin) )
+        overlays.add(ov);
+
+    for( int w = 0; w < gameMap.mapWidth; ++w )
+    {
+      if( (w+1)*tileSize < drawX || drawX + viewWidth < (w-2)*tileSize )
+        continue;
+      for( int h = 0; h < gameMap.mapHeight; ++h )
+      {
+        if( (h+1)*tileSize < drawY || drawY + viewHeight < (h-2)*tileSize )
+          continue;
+
+        XYCoord coord = new XYCoord(w,h);
+        for( GameOverlay ov : overlays )
+          if( ov.area.contains(coord) )
+            drawOverlayTile(og, ov, coord, w * tileSize - drawX, h * tileSize - drawY, tileSize);
+        
+        if( gameMap.isLocationValid(w, h) )
```

Comment:
You call `drawOverlayTile` before `isLocationValid()`?

Reply:
Hmm. I mean, it's not like we're gonna get overlays on invalid tiles.
And if we do, we're just drawing not-on-the-map. Not a huge deal.

Reply:
I mean, it should work... sure does look strange though.

Reply:
Are you saying you want it changed?

Reply:
Not saying you have to change it. I mean, you are looping over the map dimensions anyway, so `isLocationValid` should never ever return false; it would be just as valid to remove the check as it would be to move it.

Just saying it makes me question how your head works.